src/draw/ElkLayout.js

```diff
@@ -0,0 +1,344 @@
+      ...(this.elkParams.interactive ? {
+        'elk.layered.cycleBreaking.strategy': this.elkParams.cycleBreaking,
+        'elk.layered.layering.strategy': this.elkParams.layering,
+        'elk.layered.crossingMinimization.strategy': this.elkParams.crossingMinimization,
+        'elk.layered.nodePlacement.strategy': this.elkParams.nodePlacement,
+        'elk.layered.interactiveReferencePoint': this.elkParams.interactiveReferencePoint,
+      } : {}),
+      'elk.debugMode': 'true',
+
+      'elk.direction': this.elkParams.direction,
+    };
+
+    // we prepare the input in the format expected by ELK
+    const graph = {
+      id: parentNode.raw ? parentNode.raw.id : 'root', layoutOptions, children: [], edges: [],
+    };
+
+    graph.children = parentNode.children
+    // convert to the format expected by ELK
+      .map((node) => ({
+        id: node.raw.id,
+        width: node.raw.drawOption.width,
+        height: node.raw.drawOption.height,
+        x: node.raw.drawOption.x,
+        y: node.raw.drawOption.y,
+        layoutOptions: { ...layoutOptions },
+      }));
+
+    // we get the links that have source and target in the children of `node`
+    // TODO : what about cross-level links ? (ie. from a depth to another)
+    const localLinks = this.__getLinksForChildren(nodes, links, parentNode);
+
+    // convert the edges to the format expected by ELK
+    graph.edges = localLinks.map((link) => ({
+      id: `${link[0]}__${link[1]}`, sources: [link[0]], targets: [link[1]],
+    }));
+
+    // finally calling ELK
+    return ElkLayout.__elk.layout(graph);
+  }
+
+  /**
+   * Write a generated layout at a single level of hierarchy to the graph.
+   * @param {ElkNode} elkNode - previously generated layout with a single level of hierarchy.
+   * @private
+   */
+  __writeSingleDepthLayout(elkNode) {
+    const nodes = new Map(elkNode.children
+      .map((node) => [node.id, {
+        x: node.x, y: node.y,
+      }]));
+
+    this.pluginData.components.forEach((component) => {
+      if (nodes.has(component.id)) {
+        const { x, y } = nodes.get(component.id);
+
+        component.drawOption.x = x;
+        component.drawOption.y = y;
+      }
+    });
+  }
+
+  /**
+   * Compute a node's depth (ie. distance to the root).
+   * @param {NodeData} node - a node.
+   * @returns {number} depth.
+   * @private
+   */
+  __computeNodeDepth(node) {
+    if (!node.parent) {
+      // this is the root
+      return 0;
+    }
+
+    return this.__computeNodeDepth(node.parent) + 1;
+  }
+
+  /**
+   * Keep only the links that are contained within {@link nodes} children.
+   * @param {Map<string,NodeData>} nodes - all the nodes.
+   * @param {ComponentLink[]} allLinks - all the links.
+   * @param {NodeData} parentNode - node whose we want the inner links.
+   * @returns {[string,string][]} inner links represented as pairs of nodes ids.
+   * @private
+   */
+  __getLinksForChildren(nodes, allLinks, parentNode) {
```

Comment:
why not something like this: 

```js
getChildrenLinks(parentNode, links) {
    const childrenIds = new Set(parentNode.children.map((node) => node.raw?.id));
    return links
      .filter(({ source, target }) => childrenIds.has(source) && childrenIds.has(target))
      .map(({ source, target }) => ({
        id: `${source}__${target}`, sources: [source], targets: [target],
      }));
}
```

Reply:
I essentially wrote the same code, but taking into account the hierarchy (which part you removed).

 Hierarchical rendering (depth-first) allows to compute the correct size for the inner components of each parent component, triggering an ELK repositioning each time.

What you wrote would mean that we provide to ELK links that are not at the same depth in the tree, or that do not have the same parent, which is not possible at the time.

That's why I used depth / `__getAncestorByDepth` (used to be `__climbToDepth`), to get links at the same level in the tree.